### PR TITLE
Initial commit of envoy-curses

### PR DESCRIPTION
**Pull request by mccv**
Commit of an alpha-quality terminal-based UI for the Envoy admin
server. Thanks to artgillespie and falun for early feedback

Signed-off-by: Mark McBride <mark@turbinelabs.io>

### CODE REVIEW

**Comment by mattklein123** [comment]
Yes. We now have JavaScript under the project umbrella. All is well. 😉 

@danielhochman can you shepherd this in?

File: envoy-curses/src/clusters_pane.js [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+      if (!this.selectedClusterName) {
+        this.selectedClusterName = c
+        this.updateAvailableStats()
+      }
+      row.push(c)
+      row.push(this.stats.getCurrentStatValue(`cluster.${c}.upstream_cx_active`))
+      row.push(this.stats.getCurrentStatValue(`cluster.${c}.upstream_rq_active`))
+      row.push(this.stats.getCurrentStatValue(`cluster.${c}.upstream_rq_total`))
+      row.push(this.stats.getCurrentStatValue(`cluster.${c}.membership_total`))
+      row.push(this.stats.getCurrentStatValue(`cluster.${c}.membership_healthy`))
```
Comment (danielhochman) [inline envoy-curses/src/clusters_pane.js]: can you add a TODO to enumerate stats from the cluster? also i think putting info on where the group of stats comes from in the readme would be helpful

Comment (danielhochman) [reply envoy-curses/src/clusters_pane.js]: ah ok i think i was reading this wrong. i found `updateAvailableStats` but it seems some stats from /stats are missing for me.

File: envoy-curses/README.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+The application uses the [blessed](https://github.com/chjj/blessed) and
+[blessed-contrib](https://github.com/yaronn/blessed-contrib) libraries to
+provide a graphical, terminal-based interface to Envoy's admin server. Currently
+it uses the `/clusters` and `/stats` endpoints. It polls for data on a
+configurable interval, keeping a ring buffer in memory for every stat (warning:
+memory is not free, and in this program, not carefully monitored).
+
+# Usage
+
+* This isn't an escape room. Quit with ctl-c.
```
Comment (danielhochman) [inline envoy-curses/README.md]: s/ctl/ctrl/

File: envoy-curses/src/clusters_pane.js (hunk 1 above)
Comment (danielhochman) [reply envoy-curses/src/clusters_pane.js]: ok finally i actually understand the interface and why it wasn't showing for me. you can ignore these comments. deleting them.

File: envoy-curses/src/clusters_pane.js [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+   * build a set of all available stats for the currently selected cluster
+   * @returns {null} nothing
+   */
+  updateAvailableStats() {
+    const hostNames = this.clusters.getHostNames(this.selectedClusterName)
+    const newStats = new Set()
+    for (let i = 0; i < hostNames.length; i++) {
+      if (!this.clusterLevelStats.has(hostNames[i])) {
+        this.clusters.getStatNames(this.selectedClusterName, hostNames[i]).forEach(s => {
+          newStats.add(s)
```
Comment (danielhochman) [inline envoy-curses/src/clusters_pane.js]: i think we can check here if the stat is not numeric and if so then hide it. the current behavior when i select an unrenderable stat (i.e. `health_flags`) is to display an updated title but not an updated the graph.

File: envoy-curses/README.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+* navigate screens with left and right arrow keys
+* on clusters screen, select clusters with up and down arrow keys. Pressing
+  enter changes stats to the currently selected cluster.
+* press `/` or `?` to bring up stats selector. Enter selects a new stat to chart.
+
+![Cluster View](clusters-view.png)
+
+## Stats
+
+* type characters to filter stat names. Use up/down arrows to select stat to
```
Comment (danielhochman) [inline envoy-curses/README.md]: do you want to call out that you can search with regex?

File: envoy-curses/src/clusters_pane.js [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+    this.screen = options.screen
+    this.style = this.options.style || Theme.style.base
+    this.clusters = options.clusters
+    this.stats = options.stats
+    this.log = options.log
+    this.chartedStat = 'rq_total'
+    this.availableStats = []
+    this.selectedClusterName = ''
+    this.tableData = {
+      headers: ['cluster', 'cx act', 'rq act', 'rq total', 'members', 'healthy'],
```
Comment (danielhochman) [inline envoy-curses/src/clusters_pane.js]: it would be cool if the currently selected stat was added to the table for all clusters. maybe a TODO?

**Comment by danielhochman** [comment]
Very cool and definitely useful.

I think it would be great to add a list of TODOs for going from alpha -> beta quality. Either in the code or in the readme.

File: envoy-curses/src/clusters_pane.js [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+      row.push(this.stats.getCurrentStatValue(`cluster.${c}.upstream_rq_total`))
+      row.push(this.stats.getCurrentStatValue(`cluster.${c}.membership_total`))
+      row.push(this.stats.getCurrentStatValue(`cluster.${c}.membership_healthy`))
+      return row
+    })
+    this.tableData.data = newTableData
+  }
+
+  /**
+   * update underlying data model for our chart. Does not actualy attach
```
Comment (danielhochman) [inline envoy-curses/src/clusters_pane.js]: s/actualy/actually/

File: envoy-curses/src/clusters_pane.js (hunk 5 above)
Comment (mccv) [reply envoy-curses/src/clusters_pane.js]: Made this change

File: envoy-curses/README.md (hunk 4 above)
Comment (mccv) [reply envoy-curses/README.md]: yup, done

File: envoy-curses/README.md (hunk 2 above)
Comment (mccv) [reply envoy-curses/README.md]: fixed

File: envoy-curses/src/clusters_pane.js (hunk 3 above)
Comment (mccv) [reply envoy-curses/src/clusters_pane.js]: fixed

File: envoy-curses/src/clusters_pane.js (hunk 6 above)
Comment (mccv) [reply envoy-curses/src/clusters_pane.js]: fixed

**Comment by mccv** [comment]
Ok, I think this all addressed. Not sure on the merge process or set of maintainers for this repo right now, but let me know how to proceed!

**Pull request merged**